File: backend/app/services/visitor_identity.py
import hashlib
import hmac
from datetime import datetime, timezone

def daily_visitor_hash(ip_address: str | None, user_agent: str | None, secret: str) -> str | None:
    if not ip_address: return None
    date = datetime.now(timezone.utc).date().isoformat()
    normalized_ua = " ".join((user_agent or "").lower().split())
    return hmac.new(secret.encode(), f"{date}:{ip_address}:{normalized_ua}".encode(), hashlib.sha256).hexdigest()

File: backend-express/backend_node/utils/visitorIdentity.js
// Equivalent of app/services/visitor_identity.py
const crypto = require('crypto');

/** Daily-rotating, privacy-preserving visitor hash used only for human clicks. */
function dailyVisitorHash(ipAddress, userAgent, secret) {
  if (!ipAddress) return null;
  const date = new Date().toISOString().slice(0, 10); // YYYY-MM-DD, UTC
  const normalizedUa = (userAgent || '').toLowerCase().trim().split(/\s+/).join(' ');
  return crypto.createHmac('sha256', secret).update(`${date}:${ipAddress}:${normalizedUa}`).digest('hex');
}

module.exports = { dailyVisitorHash };
